crates/rpc/src/traits.rs

```diff
@@ -0,0 +1,47 @@
+use core::time::Duration;
+
+use crate::api::SupervisorApiClient;
+use alloy_primitives::Log;
+use maili_protocol::ExecutingMessage;
+
+/// TODO docs.
+#[derive(thiserror::Error, Debug)]
+pub enum ExecutingMessageValidatorError {
+    /// TODO docs.
+    #[error("TODO")]
+    Todo,
+    /// TODO docs.
+    #[error("TODO: {0}")]
+    AlloySolTypesError(#[from] alloy_sol_types::Error),
+}
+
+/// TODO docs.
+#[cfg(all(feature = "jsonrpsee", feature = "client"))]
+pub trait ExecutingMessageValidator {
+    /// TODO docs.
+    type SupervisorClient: SupervisorApiClient;
+
+    /// TODO docs.
+    fn parse_messages(
+        logs: &[Log],
+    ) -> Result<Vec<ExecutingMessage>, ExecutingMessageValidatorError> {
+        logs.iter()
+            .map(|log| {
+                // TODO: should we `impl From<Log> for ExecutingMessage`?
+                // There is `impl From<Log> for MessagePayload` but I'm unsure
+                // about the relationship between `MessagePayload` and `ExecutingMessage`
+                ExecutingMessage::abi_decode(&log.data.data, true)
```

Comment:
@emhane Do you have any advice around the log decoding? The comment above this line summarizes my question(s).

If the approach here is in fact correct (`ExecutingMessage::abi_decode(&log.data.data, true)`), then should I add some logic that checks topics before attempting to decode? So as to filter out irrelevant logs rather than error out. E.G. similar idea to `pub const CONFIG_UPDATE_TOPIC: B256`.

TY!

Reply:
you can find the logic to extract the executing messages here 
https://github.com/op-rs/kona/blob/a59f643d0627320efff49f40f4803741ae9194f1/crates/interop/src/message.rs#L102-L114
not sure what the magic number 2 should be called as a constant instead @clabby ?

you're totally right that we don't want to return error, instead we just want to filter that log out if it doesn't contain an executing message